File: sdk/python/src/inferlet/context.py
"""
Context — SDK-owned facade over the runtime KV working set.

The runtime's opaque context resource has been replaced by an inferlet-owned
``kv-working-set`` plus explicit forward-pass read/write descriptors.  This
class owns the semantic metadata (buffer, token positions, sequence length, and
replay history) and mirrors the Rust SDK facade.

Single-model runtime: the working set binds to the one bound model implicitly
(``KvWorkingSet()`` takes no handle) and model metadata is reached through the
global ``wit_world.imports.model`` / ``chat`` functions.
"""

from __future__ import annotations

import json
import os
from typing import TYPE_CHECKING, Iterable

from wit_world.imports import chat as _wit_chat
from wit_world.imports import inference as _inf
from wit_world.imports import media as _media
from wit_world.imports import model as _model
from wit_world.imports import runtime as _runtime
from wit_world.imports import working_set as _ws

from . import chat as _chat
from .forward import Forward
from .generation import Generator
from .grammar import Constraint, Schema

if TYPE_CHECKING:
    from .adapter import Adapter
    from .sample import Sampler
    from .spec import Speculator


_SNAPSHOT_VERSION = 1
_SNAPSHOT_COUNTER = 0


def _snapshot_path(name: str) -> str:
    # The runtime preopens the per-instance scratch dir as `/scratch` in the
    # guest; a relative path has no matching preopen, so blobs go there.
    return f"/scratch/{name}.pie-snapshot"


def _read_manifest(name: str) -> dict:
    with open(_snapshot_path(name), encoding="utf-8") as f:
        manifest = json.load(f)
    version = manifest.get("version")
    if version != _SNAPSHOT_VERSION:
        raise RuntimeError(
            f"snapshot '{name}': version {version} unsupported "
            f"(expected {_SNAPSHOT_VERSION})"
        )
    return manifest


def _ceil_div(n: int, d: int) -> int:
    return (n + d - 1) // d


class Context:
    """High-level inference context backed by a KV working set."""

    __slots__ = (
        "_kv",
        "_page_size",
        "_buffer",
        "_pending_system",
        "_seq_len",
        "_history",
        "_snapshottable",
    )

    # ── Construction / lifecycle ──────────────────────────────────────

    def __init__(self) -> None:
        kv = _ws.KvWorkingSet()
        self._kv = kv
        self._page_size = kv.page_size()
        self._buffer: list[int] = []
        self._pending_system: str | None = None
        self._seq_len = 0
        self._history: list[int] = []
        self._snapshottable = True

    def fork(self) -> Context:
        """Fork into a new anonymous context sharing KV pages by CoW."""
        kv = self._kv.fork()
        obj = object.__new__(Context)
        obj._kv = kv
        obj._page_size = self._page_size
        obj._buffer = list(self._buffer)
        obj._pending_system = self._pending_system
        obj._seq_len = self._seq_len
        obj._history = list(self._history)
        obj._snapshottable = self._snapshottable
        return obj

    def release(self) -> None:
        """Force-release this context's working set handle."""
        self._kv.__exit__(None, None, None)

    def __enter__(self) -> Context:
        return self

    def __exit__(self, *args) -> None:
        self.release()

    def __repr__(self) -> str:
        return f"Context({id(self._kv):#x})"

    # ── Snapshots ─────────────────────────────────────────────────────

    def save(self, name: str) -> None:
        """Save this context as a replayable CPU-resident snapshot."""
        if not self._snapshottable:
            raise RuntimeError(
                "Context.save: multimodal contexts are not snapshottable in v1 "
                "(soft-token KV cannot be replayed from a token log)"
            )
        manifest = {
            "version": _SNAPSHOT_VERSION,
            "page_size": self._page_size,
            "seq_len": self._seq_len,
            "tokens": list(self._history),
            "buffer": list(self._buffer),
            "pending_system": self._pending_system,
            "cas_hashes": [],
        }
        with open(_snapshot_path(name), "w", encoding="utf-8") as f:
            json.dump(manifest, f)

    def snapshot(self) -> str:
        """Anonymous save — returns a freshly-generated snapshot name."""
        global _SNAPSHOT_COUNTER
        name = f"anon-{_runtime.instance_id()}-{_SNAPSHOT_COUNTER}"
        _SNAPSHOT_COUNTER += 1
        self.save(name)
        return name

    @classmethod
    async def open(cls, name: str) -> Context:
        """Open a saved snapshot, replaying its token log into a fresh set."""
        manifest = _read_manifest(name)
        return await cls._from_manifest(manifest)

    @classmethod
    async def take(cls, name: str) -> Context:
        """Open a saved snapshot and then best-effort delete it."""
        ctx = await cls.open(name)
        try:
            os.remove(_snapshot_path(name))
        except OSError:
            pass
        return ctx

    @staticmethod
    def delete(name: str) -> None:
        """Delete a saved snapshot by name. Missing snapshots are ignored."""
        try:
            os.remove(_snapshot_path(name))
        except OSError:
            pass

    @classmethod
    async def _from_manifest(cls, manifest: dict) -> Context:
        ctx = cls()
        tokens = list(manifest.get("tokens", []))
        if tokens:
            ctx._buffer = tokens
            await ctx.flush()
        ctx._buffer = list(manifest.get("buffer", []))
        ctx._pending_system = manifest.get("pending_system")
        return ctx

    # ── Accessors ────────────────────────────────────────────────────

    @property
    def page_size(self) -> int:
        return self._page_size

    @property
    def seq_len(self) -> int:
        """Total materialized tokens (excludes the buffer)."""
        return self._seq_len

    def buffer(self) -> list[int]:
        """Pending (buffered but not yet flushed) tokens."""
        return list(self._buffer)

    def working_set(self) -> "_ws.KvWorkingSet":
        """Escape hatch: the underlying KV working set (power users)."""
        return self._kv

    # ── Chat fillers ─────────────────────────────────────────────────

    def _flush_pending_system(self) -> None:
        if self._pending_system is not None:
            self._buffer.extend(_chat.system(self._pending_system))
            self._pending_system = None

    def _is_first_chat_fill(self) -> bool:
        return self._seq_len == 0 and not self._buffer

    def system(self, message: str) -> Context:
        """Fill a system-role message."""
        self._flush_pending_system()
        self._pending_system = message
        return self

    def user(self, message: str) -> Context:
        """Fill a user-role message."""
        if self._pending_system is not None:
            tokens = list(_wit_chat.system_user(self._pending_system, message))
            self._pending_system = None
        elif self._is_first_chat_fill():
            tokens = list(_wit_chat.first_user(message))
        else:
            tokens = _chat.user(message)
        self._buffer.extend(tokens)
        return self

    def assistant(self, message: str) -> Context:
        """Fill an assistant-role message (history replay)."""
        self._flush_pending_system()
        self._buffer.extend(_chat.assistant(message))
        return self

    def cue(self) -> Context:
        """Cue the model to generate (fills the generation header)."""
        self._flush_pending_system()
        self._buffer.extend(_chat.cue())
        return self

    def seal(self) -> Context:
        """Seal the current turn (insert stop token)."""
        self._flush_pending_system()
        self._buffer.extend(_chat.seal())
        return self

    def append(self, tokens: Iterable[int]) -> Context:
        """Append raw tokens to the buffer directly."""
        self._flush_pending_system()
        self._buffer.extend(tokens)
        return self

    # ── Sequence / page bookkeeping ──────────────────────────────────

    def truncate(self, n: int) -> None:
        """Drop the trailing ``n`` materialized tokens and free empty slots."""
        n = min(max(n, 0), self._seq_len)
        if n == 0:
            return
        self._seq_len -= n
        keep = max(0, len(self._history) - n)
        del self._history[keep:]

        live_pages = _ceil_div(self._seq_len, self._page_size)
        have = self._kv.size()
        if have > live_pages:
            try:
                self._kv.free(list(range(live_pages, have)))
            except Exception:
                pass

    def _prepare_write(self, n: int) -> tuple[int, list[int], list[int], int]:
        p = self._page_size
        first_write_page = self._seq_len // p
        total_after = self._seq_len + n
        total_pages = _ceil_div(total_after, p)
        have = self._kv.size()
        if total_pages > have:
            self._kv.alloc(total_pages - have)
        generation = self._kv.generation()
        indices = list(range(first_write_page, total_pages))
        valid_lens = [min(total_after - pg * p, p) for pg in indices]
        return generation, indices, valid_lens, first_write_page

    def _attach_kv(
        self,
        fwd: _inf.ForwardPass,
        generation: int,
        indices: list[int],
        valid_lens: list[int],
        ctx_pages: int,
    ) -> None:
        if ctx_pages > 0:
            fwd.kv_context(
                _inf.KvContext(
                    set=self._kv,
                    start=0,
                    len=ctx_pages,
                    valid_tokens=ctx_pages * self._page_size,
                )
            )
        fwd.kv_output(
            _inf.KvOutput(
                set=self._kv,
                generation=generation,
                indices=indices,
                per_page_valid_lens=valid_lens,
            )
        )

    def _attach_full_context(self, fwd: _inf.ForwardPass) -> None:
        ctx_pages = _ceil_div(self._seq_len, self._page_size)
        if ctx_pages > 0:
            fwd.kv_context(
                _inf.KvContext(
                    set=self._kv,
                    start=0,
                    len=ctx_pages,
                    valid_tokens=self._seq_len,
                )
            )

    # ── Flush / multimodal append ────────────────────────────────────

    async def flush(self) -> None:
        """Drain buffered tokens through a forward pass into KV slots."""
        self._flush_pending_system()
        if not self._buffer:
            return
        tokens = self._buffer
        self._buffer = []
        n = len(tokens)
        positions = list(range(self._seq_len, self._seq_len + n))
        generation, indices, valid_lens, ctx_pages = self._prepare_write(n)

        fwd = _inf.ForwardPass()
        self._attach_kv(fwd, generation, indices, valid_lens, ctx_pages)
        fwd.input_tokens(tokens, positions)
        await fwd.execute()

        self._history.extend(tokens)
        self._seq_len += n

    async def append_image(self, image: _media.Image) -> None:
        """Splice an encoded image/video frame into the context."""
        prefix = list(image.prefix_tokens())
        suffix = list(image.suffix_tokens())
        if prefix:
            self.append(prefix)
        await self.flush()

        num_tokens = image.token_count()
        if num_tokens == 0:
            if suffix:
                self.append(suffix)
            return

        generation, indices, valid_lens, ctx_pages = self._prepare_write(num_tokens)
        fwd = _inf.ForwardPass()
        self._attach_kv(fwd, generation, indices, valid_lens, ctx_pages)
        fwd.input_image(image, self._seq_len)
        await fwd.execute()

        self._seq_len += num_tokens
        self._snapshottable = False
        if suffix:
            self.append(suffix)

    async def append_audio(self, audio: _media.Audio) -> None:
        """Splice an encoded audio clip into the context."""
        prefix = list(audio.prefix_tokens())
        suffix = list(audio.suffix_tokens())
        if prefix:
            self.append(prefix)
        await self.flush()

        num_tokens = audio.token_count()
        if num_tokens == 0:
            if suffix:
                self.append(suffix)
            return

        generation, indices, valid_lens, ctx_pages = self._prepare_write(num_tokens)
        fwd = _inf.ForwardPass()
        self._attach_kv(fwd, generation, indices, valid_lens, ctx_pages)
        fwd.input_audio(audio, self._seq_len)
        await fwd.execute()

        self._seq_len += num_tokens
        self._snapshottable = False
        if suffix:
            self.append(suffix)

    async def append_video(self, video: _media.Video) -> None:
        """Splice a decoded video clip into the context frame-by-frame."""
        for i in range(video.frame_count()):
            secs = max(video.timestamp(i), 0.0)
            marker = f" {int(secs) // 60:02}:{int(secs) % 60:02} "
            self.append(_model.encode(marker))
            await self.append_image(video.frame(i))

    # ── Forward (single forward-pass primitive) ──────────────────────

    def forward(self) -> Forward:
        """Build a single :class:`Forward` pass."""
        self._flush_pending_system()
        return Forward(self)

    # ── Generate (multi-step loop) ───────────────────────────────────

    def generate(
        self,
        sampler: Sampler,
        *,
        max_tokens: int | None = None,
        stop: Iterable[int] | None = None,
        constrain: Schema | Constraint | list[Schema | Constraint] | None = None,
        logit_mask: list[int] | None = None,
        speculator: Speculator | None = None,
        system_speculation: bool | None = None,
        adapter: Adapter | None = None,
        zo_seed: int | None = None,
        horizon: int | None = None,
        auto_flush: bool = True,
    ) -> Generator:
        """Build a :class:`Generator` for token generation."""
        if auto_flush:
            self.cue()
            if stop is None:
                stop = _chat.stop_tokens()

        return Generator(
            self,
            sampler,
            max_tokens=max_tokens,
            stop=stop,
            constrain=constrain,
            logit_mask=logit_mask,
            speculator=speculator,
            system_speculation=system_speculation,
            adapter=adapter,
            zo_seed=zo_seed,
            horizon=horizon,
        )

File: sdk/rust/inferlet/src/context.rs
//! SDK Context — stateful wrapper over the WIT `Context` resource.
//!
//! Owns the native context handle, caches page metadata, buffers instruct
//! tokens, and exposes ergonomic fill / flush / generate methods.

mod constraint;

// Re-export submodule public types.
pub use constraint::*;

use crate::ForwardPassExt;
use crate::Result;
use crate::inference::ForwardPass;
use crate::model::Model;
use crate::sample::Sampler;
use serde_json;

/// The raw WIT context resource, re-exported for power users.
pub use crate::pie::core::context::Context as RawContext;

// Instruct WIT bindings.
use crate::pie::instruct::chat;

/// Budget-exhausting bid: the maximum per-page-per-step rent the process
/// can sustain over `μ` steps of generation without going bankrupt.
///
/// Formula:
///
/// ```text
///     bid = (B/μ + d) / (p + μ(1 + cv²) / (2s))
/// ```
///
/// where:
/// - `B` = credit balance (market wallet, unit: pages)
/// - `μ` = expected remaining steps
/// - `d` = endowment-weighted dividend per step
/// - `p` = pages currently held
/// - `s` = page_size (tokens per page)
/// - `cv²` = squared coefficient of variation of the remaining-steps
///   distribution (0 = deterministic, 1 = geometric/memoryless)
///
/// The numerator is the per-step budget available for rent (balance
/// amortized over horizon, plus incoming dividend). The denominator is
/// the *total* page-steps of rent exposure: current pages `p` held for
/// `μ` steps, plus the triangular accumulation of newly created pages.
///
/// This is the truthful bid under critical-value payments — bidding this
/// value exhausts the wallet exactly at the end of the horizon. No
/// make-cost term: forward-pass compute is billed against the token
/// wallet, not the credit wallet.
pub(crate) fn compute_bid(
    balance: f64,
    pages: f64,
    mu: f64,
    cv2: f64,
    page_size: f64,
    dividend: f64,
) -> f64 {
    let mu = mu.max(1.0);
    let numerator = balance / mu + dividend;
    let denominator = pages + mu * (1.0 + cv2) / (2.0 * page_size);
    if denominator > 0.0 {
        numerator / denominator
    } else {
        numerator
    }
}

// =============================================================================
// Context
// =============================================================================

/// High-level inference context.
///
/// Wraps the native WIT [`RawContext`] resource and provides:
/// - **Buffered instruct fills** (`system`, `user`, `cue`, …) that accumulate
///   tokens locally.
/// - **`flush()`** to drain the buffer through a forward pass and commit pages.
/// - **`generate()`** to create a [`Generator`](crate::generation::Generator) for token-by-token generation.
/// - **Cached page metadata** (`seq_len`, `committed_pages`, `working_tokens`)
///   to avoid redundant WIT host calls.
pub struct Context {
    pub(crate) inner: RawContext,
    pub(crate) model: Model,
    pub(crate) page_size: u32,
    /// SDK-side token buffer filled by instruct operations.
    pub(crate) buffer: Vec<u32>,
    /// Deferred system text, so model templates that fold system into the
    /// first user turn can render the pair correctly.
    pending_system: Option<String>,
    /// Total tokens in committed + working pages (tracked locally).
    pub(crate) seq_len: u32,
    /// Number of committed pages (tracked locally).
    pub(crate) committed_pages: u32,
    /// Number of currently reserved working pages (tracked locally).
    pub(crate) working_pages: u32,
    /// Number of tokens in working (uncommitted) pages (tracked locally).
    pub(crate) working_tokens: u32,
}

impl Context {
    // ── Constructors ────────────────────────────────────────────────

    /// Create a fresh empty context for the given model.
    pub fn new(model: &Model) -> Result<Self> {
        let inner = RawContext::create(model)?;
        Ok(Self::wrap(inner))
    }

    /// Open a saved snapshot (implicit fork — snapshot stays immutable).
    pub fn open(model: &Model, name: &str) -> Result<Self> {
        let inner = RawContext::open(model, name)?;
        Ok(Self::wrap(inner))
    }

    /// Take ownership of a saved snapshot (snapshot is deleted).
    pub fn take(model: &Model, name: &str) -> Result<Self> {
        let inner = RawContext::take(model, name)?;
        Ok(Self::wrap(inner))
    }

    /// Delete a saved snapshot by name (static — no context needed).
    pub fn delete(model: &Model, name: &str) -> Result<()> {
        RawContext::delete(model, name)
    }

    /// Wrap an existing raw context, syncing cached state from the host.
    fn wrap(inner: RawContext) -> Self {
        let page_size = inner.tokens_per_page();
        let committed_pages = inner.committed_page_count();
        let working_pages = inner.working_page_count();
        let working_tokens = inner.working_page_token_count();
        let seq_len = committed_pages * page_size + working_tokens;
        let model = inner.model();
        Self {
            inner,
            model,
            page_size,
            buffer: Vec::new(),
            pending_system: None,
            seq_len,
            committed_pages,
            working_pages,
            working_tokens,
        }
    }

    // ── Lifecycle ───────────────────────────────────────────────────

    /// Fork into a new anonymous context (working pages are copied).
    ///
    /// The forked context inherits a copy of the parent's buffered tokens.
    pub fn fork(&self) -> Result<Self> {
        let forked = self.inner.fork()?;
        let model = forked.model();
        Ok(Self {
            inner: forked,
            model,
            page_size: self.page_size,
            buffer: self.buffer.clone(),
            pending_system: self.pending_system.clone(),
            seq_len: self.seq_len,
            committed_pages: self.committed_pages,
            working_pages: self.working_pages,
            working_tokens: self.working_tokens,
        })
    }

    /// Save the context under a user-chosen name.
    pub fn save(&self, name: &str) -> Result<()> {
        self.inner.save(name)
    }

    /// Anonymous save — returns a runtime-generated snapshot name.
    pub fn snapshot(&self) -> Result<String> {
        self.inner.snapshot()
    }

    /// Force-destroy the context immediately, consuming it.
    pub fn destroy(self) {
        self.inner.destroy()
    }

    // ── Market operations ────────────────────────────────────────────

    /// Suspend this context (release pages, stop rent).
    /// Restoration is system-driven based on bid priority.
    pub fn suspend(&self) -> Result<()> {
        self.inner.suspend()
    }

    /// Override the auto-computed bid (willingness to pay per page per
    /// step). Higher bid = harder to evict, restored first under
    /// contention. Bids are bounded below by zero; the runtime refuses
    /// negative values.
    ///
    /// Most callers should NOT use this — the [`Generator`] auto-bids
    /// every step using a budget-exhausting strategy that drains the
    /// wallet over the horizon. Reach for `set_bid` only when you have a
    /// strategy of your own.
    ///
    /// Stopping forward progress when the compute budget is spent is
    /// handled on the **token wallet**, not here — calling forward
    /// passes after `tokens_remaining == 0` fails with an error,
    /// independent of any bid you set. A low bid only affects admission
    /// under contention.
    ///
    /// [`Generator`]: crate::generation::Generator
    pub fn set_bid(&self, value: f64) {
        self.inner.bid(value);
    }

    /// Mark this context as idle: drop the bid to zero so other
    /// contexts can take its pages under contention. Returns an opaque
    /// RAII guard; the truthful generation bid is restored on drop.
    ///
    /// Use across an external wait (HTTP, tool call, anything off-GPU)
    /// where holding the bid would buy you nothing:
    ///
    /// ```ignore
    /// let _idle = ctx.idle();
    /// let result = http_get(url).await?;
    /// // _idle dropped here → bid restored
    /// ```
    ///
    /// On an uncontended device the runtime charges zero rent anyway —
    /// `idle` is a no-op cost-wise but still safe to call. Under load,
    /// it yields priority to other workloads for the duration of the
    /// wait.
    pub fn idle(&self) -> Idle<'_> {
        // Snapshot the truthful bid to restore on drop. If pages == 0
        // (rare), there's nothing to bid for; default to 0.0.
        let pages = (self.committed_pages + self.working_pages) as f64;
        let saved = if pages > 0.0 {
            let balance = crate::scheduling::balance(&self.model);
            let dividend = crate::scheduling::dividend(&self.model);
            let page_size = self.page_size as f64;
            // Conservative μ = 4096; no per-generation horizon visible here.
            compute_bid(balance, pages, 4096.0, 1.0, page_size, dividend)
        } else {
            0.0
        };
        self.inner.bid(0.0);
        Idle { ctx: self, saved }
    }

    // ── Accessors (no WIT calls) ────────────────────────────────────

    /// The model this context was created with.
    pub fn model(&self) -> &Model {
        &self.model
    }

    /// Tokens per page.
    pub fn page_size(&self) -> u32 {
        self.page_size
    }

    /// Total sequence length (committed + working tokens, excluding buffer).
    pub fn seq_len(&self) -> u32 {
        self.seq_len
    }

    /// Pending (buffered but not yet flushed) tokens.
    pub fn buffer(&self) -> &[u32] {
        &self.buffer
    }

    /// Access the underlying WIT context resource (escape hatch).
    pub fn inner(&self) -> &RawContext {
        &self.inner
    }

    // ── Instruct Fillers ────────────────────────────────────────────
    //
    // Each filler delegates to the WIT free function (which only needs the
    // model for template lookup / tokenization) and appends the resulting
    // tokens to the local buffer.

    fn flush_pending_system(&mut self) {
        if let Some(system) = self.pending_system.take() {
            let tokens = chat::system(&self.model, &system);
            self.buffer.extend(tokens);
        }
    }

    fn is_first_chat_fill(&self) -> bool {
        self.seq_len == 0 && self.buffer.is_empty()
    }

    /// Fill a system message; tokens are buffered for the next `flush()`.
    pub fn system(&mut self, message: &str) -> &mut Self {
        self.flush_pending_system();
        self.pending_system = Some(message.to_string());
        self
    }

    /// Fill a user message.
    pub fn user(&mut self, message: &str) -> &mut Self {
        let tokens = match self.pending_system.take() {
            Some(system) => chat::system_user(&self.model, &system, message),
            None if self.is_first_chat_fill() => chat::first_user(&self.model, message),
            None => chat::user(&self.model, message),
        };
        self.buffer.extend(tokens);
        self
    }

    /// Fill an assistant message (for history replay).
    pub fn assistant(&mut self, message: &str) -> &mut Self {
        self.flush_pending_system();
        let tokens = chat::assistant(&self.model, message);
        self.buffer.extend(tokens);
        self
    }

    /// Cue the model to generate (fills the generation header).
    pub fn cue(&mut self) -> &mut Self {
        self.flush_pending_system();
        let tokens = chat::cue(&self.model);
        self.buffer.extend(tokens);
        self
    }

    /// Seal the current turn (insert stop token).
    pub fn seal(&mut self) -> &mut Self {
        self.flush_pending_system();
        let tokens = chat::seal(&self.model);
        self.buffer.extend(tokens);
        self
    }

    /// Append raw tokens to the buffer directly.
    pub fn append(&mut self, tokens: &[u32]) -> &mut Self {
        self.flush_pending_system();
        self.buffer.extend_from_slice(tokens);
        self
    }

    /// Register `tools` in the chat template's tool block. Each tool's
    /// metadata is wrapped in the `{name, description, parameters}` envelope
    /// the host expects, then spliced into the buffer via the model's
    /// `equip_prefix`.
    ///
    /// Use the [`#[tool]`](inferlet_macros::tool) macro to derive a `Tool`
    /// impl from a Rust async fn, or implement the trait by hand for
    /// dynamically-loaded tools.
    ///
    /// # Errors
    /// Returns the underlying schema-parse or `equip_prefix` error if a
    /// tool's `schema()` is not valid JSON, or if the model has no tool
    /// template.
    pub fn equip(&mut self, tools: &[&dyn crate::tools::Tool]) -> Result<&mut Self> {
        self.flush_pending_system();
        let envelopes: Vec<String> = tools
            .iter()
            .map(|t| {
                let parsed: serde_json::Value = serde_json::from_str(t.schema())
                    .map_err(|e| format!("tool `{}`: invalid schema: {e}", t.name()))?;
                Ok(serde_json::json!({
                    "name": t.name(),
                    "description": t.description(),
                    "parameters": parsed,
                })
                .to_string())
            })
            .collect::<Result<_>>()?;
        let prefix = crate::tools::equip_prefix(&self.model, &envelopes)?;
        self.buffer.extend_from_slice(&prefix);
        Ok(self)
    }

    /// Drop the trailing `n` tokens from the working pages and re-sync the
    /// cached page/seq counters from the host.
    ///
    /// Use after a forward pass that wrote speculative draft tokens, to
    /// roll back the rejected suffix. `n` counts only working-page tokens —
    /// pages that already committed cannot be truncated through this API.
    pub fn truncate(&mut self, n: u32) {
        if n == 0 {
            return;
        }
        let keep = self.inner.working_page_token_count().saturating_sub(n);
        self.inner.truncate_working_page_tokens(keep);
        // Re-sync from the host: truncation can release pages, and the
        // safe thing is to read the authoritative counts back.
        self.committed_pages = self.inner.committed_page_count();
        self.working_pages = self.inner.working_page_count();
        self.working_tokens = self.inner.working_page_token_count();
        self.seq_len = self.committed_pages * self.page_size + self.working_tokens;
    }

    // ── Flush ───────────────────────────────────────────────────────

    /// Drain the buffered tokens through a forward pass and commit pages.
    ///
    /// After flush, the buffer is empty and `seq_len` reflects all
    /// consumed tokens.
    pub async fn flush(&mut self) -> Result<()> {
        self.flush_pending_system();
        if self.buffer.is_empty() {
            return Ok(());
        }

        let tokens = std::mem::take(&mut self.buffer);
        let num_tokens = tokens.len() as u32;

        // Reserve additional pages if we need more than currently allocated.
        let total_tokens_after = self.working_tokens + num_tokens;
        let pages_needed = (total_tokens_after + self.page_size - 1) / self.page_size;
        let additional = pages_needed.saturating_sub(self.working_pages);
        if additional > 0 {
            self.inner
                .reserve_working_pages(additional)
                .map_err(|e| format!("Failed to reserve pages: {}", e))?;
            self.working_pages = pages_needed;
        }

        // Build and execute a forward pass.
        let pass = ForwardPass::new(&self.model);
        pass.context(&self.inner);

        let positions: Vec<u32> = (self.seq_len..self.seq_len + num_tokens).collect();
        pass.input_tokens(&tokens, &positions);

        pass.execute_async()
            .await
            .map_err(|e| format!("Forward pass failed: {}", e))?;

        // Commit full pages.
        let new_working = self.working_tokens + num_tokens;
        let pages_to_commit = new_working / self.page_size;
        if pages_to_commit > 0 {
            self.inner
                .commit_working_pages(pages_to_commit)
                .map_err(|e| format!("Failed to commit pages: {}", e))?;
        }

        // Update cached state.
        self.committed_pages += pages_to_commit;
        self.working_pages -= pages_to_commit;
        self.working_tokens = new_working % self.page_size;
        self.seq_len += num_tokens;

        Ok(())
    }

    // ── Pass ────────────────────────────────────────────────────────

    /// Build a single [`Forward`](crate::forward::Forward) — a forward pass with
    /// automatic page reservation, position derivation, and post-execute
    /// commit. Use for prefill, scoring, custom decode loops, and anywhere
    /// the [`generate`](Self::generate) loop is too high-level.
    pub fn forward(&mut self) -> crate::forward::Forward<'_> {
        self.flush_pending_system();
        crate::forward::Forward::new(self)
    }

    // ── Generate ────────────────────────────────────────────────────

    /// Build a [`Generator`](crate::generation::Generator) — the
    /// multi-step token-generation state machine. Any tokens already in
    /// the buffer (from `system / user / cue / …`) are drained on the
    /// first step.
    pub fn generate(&mut self, sampler: Sampler) -> crate::generation::Generator<'_> {
        self.flush_pending_system();
        crate::generation::Generator::new(self, sampler)
    }
}

// =============================================================================
// Idle — RAII guard returned by Context::idle()
// =============================================================================

/// Opaque RAII guard. Created by [`Context::idle`]. Drops the bid to
/// zero for the duration of the guard's lifetime; restores the
/// truthful bid on drop.
pub struct Idle<'a> {
    ctx: &'a Context,
    saved: f64,
}

impl<'a> Drop for Idle<'a> {
    fn drop(&mut self) {
        self.ctx.inner.bid(self.saved);
    }
}
